### PR TITLE
doc: filter chain manager

### LINKED ISSUE #12572: filter_chain_match difference between {} and null is confusing/under documented
If I have config (example 1):
```yaml
match:
  app_protocols: [h2]
match: {}
```

All h2 traffic matches filter 1, and everything else matches filter 2. If I then change this to (example 2)
```yaml
match:
  app_protocols: [h2]
  transport_protocol: tls
match: {}
```

No traffic matches filter 2. I would expect filter 2 to still catch all unmatched traffic. It seems in order to do this, I need to specify: (example 3)
```yaml
match:
  app_protocols: [h2]
  transport_protocol: tls
match: null
```

I think this contradicts the docs a bit. Transport protocol says "If non-empty, a transport protocol to consider when determining a filter chain match.". In the example 2, I have set transport_protocol to `""`, yet it still seems to be considering transport protocol.

Even more contradicting is that changing one FCM (seems to) impacts the matching logic of another FCM! My expectation would be that a match config is independent. If I have `{}` it would always match the same inputs (from my reading, match everything, but at least should be consistent), regardless of other FCMs - other FCMs may have higher priority, but not actually prevent this one from matching.

cc @lizan @rshriram

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
Describe filter chain match decision through the sieve method.

Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
Fix #12572
[Optional Deprecated:]

### CODE REVIEW

File: api/envoy/config/listener/v3/listener_components.proto [hunk 1]
```diff
@@ -64,6 +64,18 @@ message Filter {
 // ``www.example.com``, then ``*.example.com``, then ``*.com``, then any filter
 // chain without ``server_names`` requirements).
 //
+// Yet another approach to reason about the filter chain matches.
```
Comment (htuch) [inline api/envoy/config/listener/v3/listener_components.proto:67]: To me, the sieve explanation is less intuitive than an ordered walk through the criteria, but this is subjective. I'll defer to other @envoyproxy/api-shepherds on whether they think this adds value.

File: api/envoy/config/listener/v3/listener_components.proto [hunk 2]
```diff
@@ -64,6 +64,18 @@ message Filter {
...
+// Suppose you have N filter chains in hand. Sieve the filter chain set in the above 8 rounds.
+// In each round, filter chains matching the attributes best survives to the next round.
+// The listener guarantees at most 1 filter chain left after all the rounds.
+//
+// Example:
+//
+// For destination port, filter chains specifying the destination port of incoming traffic are the
```
Comment (howardjohn) [inline api/envoy/config/listener/v3/listener_components.proto:74]: So if I have 1 FC with a destination port, and another without destination port set, the latter will never match? And if I do want "everything but destination port from FC1" I need to make 65535 FCs? Or am I reading wrong

**Comment by mattklein123** [comment]
Note this also fixed https://github.com/envoyproxy/envoy/issues/11088. @lambdai can you please also document the "smart filter chain updates" functionality while you are working on this?

**Comment by lizan** [comment]
It would be also nice to document how to configure a "catch-all" filter chain match. That's what we were confused in #12572

File: api/envoy/config/listener/v3/listener_components.proto (hunk 2 above)
Comment (lambdai) [reply api/envoy/config/listener/v3/listener_components.proto:74]: More concrete example, suppose listener has 4 filter chains
FC1 {dest port 80}
FC 2{dest port 90}
FC 3{dest port: null}
FC 4{dest port: null, ip = 10.0.0.1}

For incoming traffic which has  
a) dest port = 80
    {FC1} is the best match
b) dest port = 90
   {FC2} is the best match
c) dst port = 100
   {FC3, FC4} is the best match

**Comment by lambdai** [comment]
@mattklein123 ah, the debt. Sure

@lizan I haven't implement the authentic catch all in istio... But yes, I can add the document here first

**Comment by lambdai** [comment]
While I am writing the pseudo code to calculate what filter chains are needed to achieve a fallback, I changed my mind and agree that envoy should add a fallback filter chain.
Otherwise the control plane like istio need to create it anyway. And the cost is huge: we might need to duplicate the filter chains many many times. All of them share the same filters but with different filter chain match. It is more reasonable to compress the fallback filter chain set into a magical "fallback filter chain".
It could be implemented in envoy as an extra branch to see if there is fallback filter chain beside [here
](https://github.com/envoyproxy/envoy/blob/933e267db60961e76549ae6e65ac4872cb78db4d/source/server/connection_handler_impl.cc#L394)

**Comment by mattklein123** [comment]
@lambdai sorry are we reviewing this or are we now doing a code change to add a feature and then doc that?

/wait-any

**Pull request closed without merging**